File: core-ClientUnity - Copy/Assets/Scripts/ModelingData.cs
//using System;
//using System.Collections;
//
//namespace ServerCommunication
//{
//    public enum DataType
//    {
//        DT_BASE             = 100,
//        DT_POINT            = 101,
//        DT_POLYLINE         = 102,
//        DT_PATH             = 103,
//        DT_MODELRAW         = 104,
//        DT_MODELMESH        = 105,
//    };
//
//    public enum ActuratorCommand
//    {
//        AC_EMPTY = 0,
//        AC_UNDO = 1,
//        AC_CUR = 2,     // Tip position
//        AC_DOWN = 3,        // Tip down position
//        AC_UP = 4,      // Tip up position
//    };
//
//    public class Modeling
//    {
//        public DataType type;
//        public Int32 count;
//        public Int32 id;
//        public ActuratorCommand command;
//
//        public virtual byte[] ToByteArray()
//        {
//            count = byte_size();
//
//            byte[] _type = BitConverter.GetBytes((Int32)type);
//            byte[] _count = BitConverter.GetBytes(count);
//
//            byte[] _id = BitConverter.GetBytes(id);
//            byte[] _command = BitConverter.GetBytes((Int32)command);
//
//            byte[] ret = new byte[byte_size()];
//
//            _type.CopyTo(ret, 0);
//            _count.CopyTo(ret, _type.Length);
//            _id.CopyTo(ret, _type.Length + _count.Length);
//            _command.CopyTo(ret, _count.Length + _id.Length + _type.Length);
//
//            return ret;
//        }
//
//        public virtual Int32 byte_size()
//        {
//            return sizeof(DataType) + sizeof(Int32) + sizeof(Int32) + sizeof(ActuratorCommand);
//        }
//
//        public Modeling()
//        {
//            id = -1;
//            type = DataType.DT_BASE;
//            command = ActuratorCommand.AC_EMPTY;
//        }
//
//        public Modeling(byte[] data, int index = 0)
//        {
//            type = (DataType)BitConverter.ToInt32(data, index);
//            count = BitConverter.ToInt32(data, index + sizeof(DataType));
//            id = BitConverter.ToInt32(data, index + sizeof(Int32) + sizeof(DataType));
//            command = (ActuratorCommand)BitConverter.ToInt32(data, index + sizeof(Int32) + sizeof(Int32) + sizeof(DataType));
//        }
//    };
//
//    public class Point : Modeling
//    {
//        public float x;
//        public float y;
//        public float z;
//
//        public override byte[] ToByteArray()
//        {
//            byte[] _base = base.ToByteArray();
//
//            byte[] _x = BitConverter.GetBytes(x);
//            byte[] _y = BitConverter.GetBytes(y);
//            byte[] _z = BitConverter.GetBytes(z);
//
//            byte[] ret = new byte[byte_size()];
//
//            _base.CopyTo(ret, 0);
//            _x.CopyTo(ret, _base.Length);
//            _y.CopyTo(ret, _base.Length + _x.Length);
//            _z.CopyTo(ret, _base.Length + _x.Length + _y.Length);
//
//            return ret;
//        }
//
//        public override Int32 byte_size()
//        {
//            return base.byte_size() + sizeof(float) * 3;
//        }
//
//        public Point()
//        {
//            type = DataType.DT_POINT; command = ActuratorCommand.AC_EMPTY;
//        }
//
//        public Point(float _x, float _y, float _z)
//        {
//            type = DataType.DT_POINT;
//            command = ActuratorCommand.AC_EMPTY;
//            x = _x; y = _y; z = _z;
//        }
//
//        public Point(byte[] data, int index = 0) : base(data, index)
//        {
//            x = BitConverter.ToSingle(data, index + base.byte_size());
//            y = BitConverter.ToSingle(data, index + base.byte_size() + sizeof(float));
//            z = BitConverter.ToSingle(data, index + base.byte_size() + sizeof(float) * 2);
//        }
//    };
//
//    public class Polyline : Modeling
//    {
//        public ArrayList points = new ArrayList();
//
//        public override byte[] ToByteArray()
//        {
//            byte[] _base = base.ToByteArray();
//
//            byte[] ret = new byte[byte_size()];
//
//            _base.CopyTo(ret, 0);
//            int index = _base.Length;
//
//            int count = points.Count;
//            byte[] len = BitConverter.GetBytes((Int32)count);
//            len.CopyTo(ret, index);
//            index += len.Length;
//
//            for(int i = 0; i < points.Count; i++)
//            {
//                byte[] pt = (points[i] as Point).ToByteArray();
//                pt.CopyTo(ret, index);
//
//                index += pt.Length;
//            }
//
//            return ret;
//        }
//
//        public override Int32 byte_size()
//        {
//            int index = base.byte_size() + sizeof(Int32);
//            for (int i = 0; i < points.Count; i++)
//                index += (points[i] as Point).byte_size();
//
//            return index;
//        }
//
//        public Polyline()
//        {
//            type = DataType.DT_POLYLINE; command = ActuratorCommand.AC_EMPTY;
//        }
//
//        public Polyline(byte[] data, int start = 0) : base(data, start)
//        {
//            int index = start + base.byte_size();
//
//            int count = BitConverter.ToInt32(data, index);
//            index += sizeof(Int32);
//
//            for (int i = 0; i < count; i++)
//            {
//                Point pt = new Point(data, index);
//
//                points.Add(pt);
//
//                index += pt.byte_size();
//            }
//        }
//    };
//
//    public class Path : Modeling
//    {
//        public ArrayList points = new ArrayList();
//
//        public override byte[] ToByteArray()
//        {
//            byte[] _base = base.ToByteArray();
//
//            byte[] ret = new byte[byte_size()];
//
//            _base.CopyTo(ret, 0);
//            int index = _base.Length;
//
//            int count = points.Count;
//            byte[] len = BitConverter.GetBytes((Int32)count);
//            len.CopyTo(ret, index);
//            index += len.Length;
//
//            for (int i = 0; i < points.Count; i++)
//            {
//                byte[] pt = (points[i] as Point).ToByteArray();
//                pt.CopyTo(ret, index);
//
//                index += pt.Length;
//            }
//
//            return ret;
//        }
//
//        public override Int32 byte_size()
//        {
//            int index = base.byte_size() + sizeof(Int32);
//            for (int i = 0; i < points.Count; i++)
//                index += (points[i] as Point).byte_size();
//
//            return index;
//        }
//
//        public Path()
//        {
//            type = DataType.DT_PATH; command = ActuratorCommand.AC_EMPTY;
//        }
//
//        public Path(byte[] data, int start = 0) : base(data, start)
//        {
//            int index = start + base.byte_size();
//
//            int count = BitConverter.ToInt32(data, index);
//            index += sizeof(Int32);
//
//            for (int i = 0; i < count; i++)
//            {
//                Point pt = new Point(data, index);
//
//                points.Add(pt);
//
//                index += pt.byte_size();
//            }
//        }
//    };
//
//    public class ModelRaw : Modeling
//    {
//        // Modeling data
//        public Polyline polyline = new Polyline();
//        public Path         path = new Path();
//
//        public override byte[] ToByteArray()
//        {
//            byte[] _base = base.ToByteArray();
//
//            byte[] ret = new byte[byte_size()];
//
//            _base.CopyTo(ret, 0);
//            int index = _base.Length;
//
//            // polyline --------------------------------
//            byte[] _polyline = polyline.ToByteArray();
//
//            int len = _polyline.Length;
//            byte[] l_p = BitConverter.GetBytes((Int32)len);
//            l_p.CopyTo(ret, index);
//            index += l_p.Length;
//
//            _polyline.CopyTo(ret, index);
//            index += _polyline.Length;
//
//            // path ------------------------------------
//            byte[] _path = path.ToByteArray();
//
//            len = _path.Length;
//            l_p = BitConverter.GetBytes((Int32)len);
//            l_p.CopyTo(ret, index);
//            index += l_p.Length;
//
//            _path.CopyTo(ret, index);
//
//            return ret;
//        }
//
//        public override Int32 byte_size()
//        {
//            int index = base.byte_size();
//
//            index += sizeof(Int32);
//            index += polyline.byte_size();
//            index += sizeof(Int32);
//            index += path.byte_size();
//
//            return index;
//        }
//
//        public bool isRevolution()
//        {
//            if (path.points.Count == 0)
//                return false;
//
//            float n = 0;
//            for (int i = 0; i < path.points.Count; i++)
//            {
//                Point pt = path.points[i] as Point;
//                n += pt.z;
//            }
//
//            if (n == 0)
//                return true;
//            else
//                return false;
//        }
//
//        public ModelRaw()
//        {
//            type = DataType.DT_MODELRAW; command = ActuratorCommand.AC_EMPTY;
//        }
//
//        public ModelRaw(byte[] data, int start = 0) : base(data, start)
//        {
//            int index = base.byte_size();
//
//            // polyline -----------------------------------------------------
//            polyline = new Polyline(data, index);
//            index += polyline.byte_size();
//
//            // path ---------------------------------------------------------
//            path = new Path(data, index);
//            index += path.byte_size();
//        }
//    };
//
//    public class ModelMesh : Modeling
//    {
//        // Display data
//        public ArrayList points     = new ArrayList();             // x, y, z, x, y, z ...
//        public ArrayList normals    = new ArrayList();            // nx, ny, nz, nx, ny, nz ...
//        public ArrayList triangles  = new ArrayList();          // ix, iy, iz, ix, iy, iz ...
//        public ArrayList domains    = new ArrayList();            // ic, ic, ic, ic ...
//
//        public override byte[] ToByteArray()
//        {
//            byte[] _base = base.ToByteArray();
//
//            byte[] ret = new byte[byte_size()];
//
//            _base.CopyTo(ret, 0);
//            int index = _base.Length;
//
//            // points -------------------------------
//            int len = points.Count;
//            byte[] l_p = BitConverter.GetBytes((Int32)len);
//            l_p.CopyTo(ret, index);
//            index += sizeof(Int32);
//
//            for (int i = 0; i < points.Count; i++)
//            {
//                byte[] val = BitConverter.GetBytes((float)points[i]);
//                val.CopyTo(ret, index);
//
//                index += val.Length;
//            }
//
//            // normals ------------------------------
//            len = normals.Count;
//            l_p = BitConverter.GetBytes((Int32)len);
//            l_p.CopyTo(ret, index);
//            index += sizeof(Int32);
//
//            for (int i = 0; i < normals.Count; i++)
//            {
//                byte[] val = BitConverter.GetBytes((float)normals[i]);
//                val.CopyTo(ret, index);
//
//                index += val.Length;
//            }
//
//            // triangles --------------------------------
//            len = triangles.Count;
//            l_p = BitConverter.GetBytes((Int32)len);
//            l_p.CopyTo(ret, index);
//            index += sizeof(Int32);
//
//            for (int i = 0; i < triangles.Count; i++)
//            {
//                byte[] val = BitConverter.GetBytes((Int32)triangles[i]);
//                val.CopyTo(ret, index);
//
//                index += val.Length;
//            }
//
//            // domains --------------------------------
//            len = domains.Count;
//            l_p = BitConverter.GetBytes((Int32)len);
//            l_p.CopyTo(ret, index);
//            index += sizeof(Int32);
//
//            for (int i = 0; i < domains.Count; i++)
//            {
//                byte[] val = BitConverter.GetBytes((Int32)domains[i]);
//                val.CopyTo(ret, index);
//
//                index += val.Length;
//            }
//
//            return ret;
//        }
//
//        public override Int32 byte_size()
//        {
//            int index = base.byte_size();
//
//            index += sizeof(Int32);
//            index += points.Count * sizeof(float);
//            index += sizeof(Int32);
//            index += normals.Count * sizeof(float);
//            index += sizeof(Int32);
//            index += triangles.Count * sizeof(Int32);
//            index += sizeof(Int32);
//            index += domains.Count * sizeof(Int32);
//
//            return index;
//        }
//
//        public void AddTriangle(float[] point, float[] normal)
//        {
//            int index = points.Count / 3;
//
//            for (int i = 0; i < 9; i++)
//                points.Add(points[i]);
//
//            for (int i = 0; i < 3; i++)
//                normals.Add(normal[i]);
//
//            triangles.Add(index + 1);
//        }
//
//        public ModelMesh()
//        {
//            type = DataType.DT_MODELMESH; command = ActuratorCommand.AC_EMPTY;
//        }
//
//        public ModelMesh(byte[] data, int start = 0) : base(data, start)
//        {
//            int index = base.byte_size();
//
//            // points -----------------------------------------------------
//            int len = BitConverter.ToInt32(data, index);
//            index += sizeof(Int32);
//
//            for (int i = 0; i < len; i++)
//            {
//                float val = BitConverter.ToSingle(data, index);
//                points.Add(val);
//                index += sizeof(float);
//            }
//
//            // normals -----------------------------------------------------
//            len = BitConverter.ToInt32(data, index);
//            index += sizeof(Int32);
//
//            for (int i = 0; i < len; i++)
//            {
//                float val = BitConverter.ToSingle(data, index);
//                normals.Add(val);
//                index += sizeof(float);
//            }
//
//            // triangles -----------------------------------------------------
//            len = BitConverter.ToInt32(data, index);
//            index += sizeof(Int32);
//
//            for (int i = 0; i < len; i++)
//            {
//                Int32 val = BitConverter.ToInt32(data, index);
//                triangles.Add(val);
//                index += sizeof(float);
//            }
//
//            // domains ------------------------------------------------------
//            len = BitConverter.ToInt32(data, index);
//            index += sizeof(Int32);
//
//            for (int i = 0; i < len; i++)
//            {
//                Int32 val = BitConverter.ToInt32(data, index);
//                domains.Add(val);
//                index += sizeof(float);
//            }
//        }
//    };
//}

File: core-ServerPC-image/DataBase/ModelingData.cpp
#include "ModelingData.h"

#ifdef ANDROID
#include <string.h>

#define NULL 0
#endif

// **********************************************************************************************
// modleing
Modeling::Modeling() : id(-1), type(DT_BASE), command(AC_EMPTY)
{
};

Modeling::Modeling(char buf[], int index)
{
	// get value of type
	int pos = index, len = sizeof(DataType);
	memcpy_s(&type, len, &buf[pos], len);
	pos += len;

	// get value of count
	len = sizeof(int);
	memcpy_s(&count, len, &buf[pos], len);
	pos += len;

	// get value of type
	len = sizeof(ID);
	memcpy_s(&id, len, &buf[pos], len);
	pos += len;

	// get value of command
	len = sizeof(ActuratorCommand);
	memcpy_s(&command, len, &buf[pos], len);
}

Modeling::~Modeling()
{

}

int Modeling::base_byte_size()
{
	return sizeof(int) + sizeof(ID) + sizeof(DataType) + sizeof(ActuratorCommand);
}

int Modeling::byte_size()
{
	return base_byte_size();
}

void Modeling::ToCharArray(char buf[], int index)
{
	// copy type to buf
	size_t pos = index, len = sizeof(DataType);
	memcpy_s(&(buf[pos]), len, &type, len);
	pos += len;

	// copy byte count of the object
	int count = byte_size();
	len = sizeof(int);
	memcpy_s(&(buf[pos]), len, &count, len);
	pos += len;

	// copy id to buf
	len = sizeof(ID);
	memcpy_s(&(buf[pos]), len, &id, len);
	pos += len;

	// copy command to buf
	len = sizeof(ActuratorCommand);
	memcpy_s(&(buf[pos]), len, &command, len);
}

// **********************************************************************************************
// MPoint
MPoint::MPoint()
{
	type = DT_POINT;
	command = AC_EMPTY;
};

MPoint::MPoint(float _x, float _y, float _z) : x(_x), y(_y), z(_z)
{
	type = DT_POINT;
	command = AC_EMPTY;
};

MPoint::MPoint(char buf[], int index) : Modeling(buf, index)
{
	// get value of x
	int pos = index + base_byte_size(), len = sizeof(float);
	memcpy_s(&x, len, &buf[pos], len);
	pos += len;

	// get value of y
	len = sizeof(float);
	memcpy_s(&y, len, &buf[pos], len);
	pos += len;

	// get value of z
	len = sizeof(float);
	memcpy_s(&z, len, &buf[pos], len);
	pos += len;
}

MPoint::~MPoint()
{

}

void MPoint::ToCharArray(char buf[], int index)
{
	// copy base to buf
	Modeling::ToCharArray(buf, index);

	// copy x to buf
	int pos = index + base_byte_size(), len = sizeof(float);
	memcpy_s(&buf[pos], len, &x, len);
	pos += len;

	// copy y to buf
	len = sizeof(float);
	memcpy_s(&buf[pos], len, &y, len);
	pos += len;

	// copy z to buf
	len = sizeof(float);
	memcpy_s(&buf[pos], len, &z, len);
	pos += len;
}

int MPoint::byte_size()
{
	return base_byte_size() + sizeof(float) * 3;
}

// **********************************************************************************************
// MPolyline
MPolyline::MPolyline()
{
	type = DT_POLYLINE;
	command = AC_EMPTY;

}

MPolyline::MPolyline(char buf[], int index) : Modeling(buf, index)
{
	// find count of points
	int pos = index + base_byte_size(), len = sizeof(int), count;
	memcpy_s(&count, len, &buf[pos], len);
	pos += len;

	// get value of points
	for (int i = 0; i < count; i++)
	{
		MPoint pt(buf, pos);
		points.push_back(pt);
		pos += pt.byte_size();
	}
}

MPolyline::~MPolyline()
{
};

void MPolyline::ToCharArray(char buf[], int index)
{
	// copy base to buf
	Modeling::ToCharArray(buf, index);

	// copy count of points to buf
	int count = (int)points.size();
	int pos = index + base_byte_size(), len = sizeof(int);
	memcpy_s(&buf[pos], len, &count, len);
	pos += len;

	// copy each point
	for (size_t i = 0; i < points.size(); i++)
	{
		points[i].ToCharArray(buf, pos);
		pos += points[i].byte_size();
	}
}

int MPolyline::byte_size()
{
	int len = base_byte_size() + sizeof(int);
	for (size_t i = 0; i < points.size(); i++)
		len += points[i].byte_size();

	return len;
}

// **********************************************************************************************
// Path
MPath::MPath()
{
	type = DT_PATH;
	command = AC_EMPTY;
}

MPath::MPath(char buf[], int index) : Modeling(buf, index)
{
	// find count of points
	int pos = index + base_byte_size(), len = sizeof(int);
	memcpy_s(&count, len, &buf[pos], len);
	pos += len;

	// get value of points
	for (int i = 0; i < count; i++)
	{
		MPoint pt(buf, pos);
		points.push_back(pt);
		pos += pt.byte_size();
	}
}

MPath::~MPath()
{
};

void MPath::ToCharArray(char buf[], int index)
{
	// copy base to buf
	Modeling::ToCharArray(buf, index);

	// copy count of points to buf
	int count = (int)points.size();
	int pos = index + base_byte_size(), len = sizeof(int);
	memcpy_s(&buf[pos], len, &count, len);
	pos += len;

	// copy each point
	for (size_t i = 0; i < points.size(); i++)
	{
		points[i].ToCharArray(buf, pos);
		pos += points[i].byte_size();
	}
}

int MPath::byte_size()
{
	int len = base_byte_size() + sizeof(int);
	for (size_t i = 0; i < points.size(); i++)
		len += points[i].byte_size();

	return len;
}

// **********************************************************************************************
// Model3D
ModelRaw::ModelRaw()
{
	type = DT_MODELRAW;
	command = AC_EMPTY;
};

ModelRaw::ModelRaw(char buf[], int index) : Modeling(buf, index)
{
	// get value of polyline
	int pos = index + base_byte_size();
	polyline = MPolyline(buf, pos);
	pos += polyline.byte_size();

	// get value of path
	path = MPath(buf, pos);
	pos += path.byte_size();
}

ModelRaw::~ModelRaw()
{

}

void ModelRaw::ToCharArray(char buf[], int index)
{
	// copy base to buf
	Modeling::ToCharArray(buf, index);

	// copy polyline to buf
	int pos = index + base_byte_size();
	polyline.ToCharArray(buf, pos);
	pos += polyline.byte_size();

	// copy path to buf
	path.ToCharArray(buf, pos);
	pos += path.byte_size();
}

int ModelRaw::byte_size()
{
	int len = base_byte_size();

	len += polyline.byte_size();

	len += path.byte_size();

	return len;
}

bool ModelRaw::isRevolution()
{
	if (path.points.size() == 0)
		return false;

	float n = 0;
	for (size_t i = 0; i < path.points.size(); i++)
	{
		n += path.points[i].z;
	}

	if (n == 0)
		return true;
	else
		return false;
}

// **********************************************************************************************
// ModelMesh
ModelMesh::ModelMesh()
{
	type = DT_MODELMESH;
	command = AC_EMPTY;
};

ModelMesh::ModelMesh(char buf[], int index) : Modeling(buf, index)
{
	// get value of points
	int pos = index + base_byte_size(), len = sizeof(int), count;
	memcpy_s(&count, len, &buf[pos], len);
	pos += len;

	if (count > 0)
	{
		points.resize(count);

		len = count * sizeof(float);
		memcpy_s(&points[0], len, &buf[pos], len);
		pos += len;
	}

	// get value of normals
	len = sizeof(int);
	memcpy_s(&count, len, &buf[pos], len);
	pos += len;

	if (count > 0)
	{
		normals.resize(count);

		len = count * sizeof(float);
		memcpy_s(&normals[0], len, &buf[pos], len);
		pos += len;
	}

	// get value of triangles
	len = sizeof(int);
	memcpy_s(&count, len, &buf[pos], len);
	pos += len;

	if (count > 0)
	{
		triangles.resize(count);

		len = count * sizeof(int);
		memcpy_s(&triangles[0], len, &buf[pos], len);
		pos += len;
	}

	// get value of domains
	len = sizeof(int);
	memcpy_s(&count, len, &buf[pos], len);
	pos += len;

	if (count > 0)
	{
		domains.resize(count);

		len = count * sizeof(int);
		memcpy_s(&domains[0], len, &buf[pos], len);
		pos += len;
	}
}

ModelMesh::~ModelMesh()
{

}

void ModelMesh::ToCharArray(char buf[], int index)
{
	// copy base to buf
	Modeling::ToCharArray(buf, index);

	// copy points to buf
	int count = (int)points.size();
	int pos = index + base_byte_size(), len = sizeof(int);
	memcpy_s(&buf[pos], len, &count, len);
	pos += len;

	if (count > 0)
	{
		len = count * sizeof(float);
		memcpy_s(&buf[pos], len, &points[0], len);
		pos += len;
	}

	// copy normals to buf
	count = (int)normals.size();
	len = sizeof(int);
	memcpy_s(&buf[pos], len, &count, len);
	pos += len;

	if (count > 0)
	{
		len = count * sizeof(float);
		memcpy_s(&buf[pos], len, &normals[0], len);
		pos += len;
	}

	// copy triangles to buf
	count = (int)triangles.size();
	len = sizeof(int);
	memcpy_s(&buf[pos], len, &count, len);
	pos += len;

	if (count > 0)
	{
		len = count * sizeof(int);
		memcpy_s(&buf[pos], len, &triangles[0], len);
		pos += len;
	}

	// copy domains to buf
	count = (int)domains.size();
	len = sizeof(int);
	memcpy_s(&buf[pos], len, &count, len);
	pos += len;

	if (count > 0)
	{
		len = count * sizeof(int);
		memcpy_s(&buf[pos], len, &domains[0], len);
		pos += len;
	}
}

int ModelMesh::byte_size()
{
	int len = base_byte_size();

	len += sizeof(int);
	len += sizeof(float) * (int)points.size();

	len += sizeof(int);
	len += sizeof(float) * (int)normals.size();

	len += sizeof(int);
	len += sizeof(int) * (int)triangles.size();

	len += sizeof(int);
	len += sizeof(int) * (int)domains.size();

	return len;
}

void ModelMesh::AppendTriangle(float pt[9], float normal[3])
{
	size_t index = points.size() / 9;

	for (int i = 0; i < 9; i++) {
		points.push_back(pt[i]);
	}

	for (int i = 0; i < 3; i++) {
		normals.push_back(normal[i]);
	}

	triangles.push_back((int)index);
}

void ModelMesh::AppendDomain(int count)
{
	domains.push_back(count);
}
